src/librustc_typeck/check/mod.rs

```diff
@@ -3357,6 +3373,36 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         vec![self.tcx.types.err; len]
     }
 
+    fn point_at_arg_instead_of_call_if_possible(
+        &self,
+        errors: &mut Vec<traits::FulfillmentError<'_>>,
+        final_arg_types: &[(usize, Ty<'tcx>)],
+        call_sp: Span,
+        args: &'tcx [hir::Expr],
+    ) {
+        if !call_sp.desugaring_kind().is_some() {
+            // We *do not* do this for desugared call spans to keep good diagnostics when involving
+            // the `?` operator.
+            for error in errors {
+                if let ty::Predicate::Trait(predicate) = error.obligation.predicate {
+                    let mut referenced_in = vec![];
```

Comment:
Try this? (Avoids the allocation)
```rust
let referenced_in = final_arg_types
    .iter()
    .flat_map(|(i, ty)| {
        let ty = self.resolve_vars_if_possible(ty);
        ty.walk()
           .filter(|&ty| ty == predicate.skip_binder().self_ty())
           .map(|_| *i)
    });

if let Some(ref_in) = referenced_in.next() {
    ...
}
```

Reply:
We want to make sure there's only one argument that can fit the obligation failure. Because of that the allocation will be needed, but only happens in the error path.

Reply:
You can use `if let (Some(ref_in), None) = (referenced_in.next(), referenced_in.next())` to check that nothing follows the first argument.